### DISCUSSION TITLE
Having trouble with the media upload process... somewhere.

### DISCUSSION THREAD

**Discussion by karsteny** [redouane59/twittered#385] [Q&A]
Hi! I'm quite new to using this (and also still pretty inexperienced in Java programming itself), so maybe it's just a simple mistake. The issue is that I cannot get any media (or at least the images I tried) to upload. I'm not too sure if its an issue with how I am reading the file to the Byte array, or syntax, but it always results in either `javax.net.ssl.SSLException: Received fatal alert: record_overflow` or `java.io.IOException: Error writing to server`, depending on the image. 
Here is the code:  
```java
File file = new File("cover.jpg");
byte[] fileContent = Files.readAllBytes(file.toPath());
UploadMediaResponse media = client.uploadMedia("cover.jpg", fileContent, MediaCategory.TWEET_IMAGE);
```
The files are all <5MB and they appear to be read in correctly if I list out the contents of the byte array. The client instance itself will work just fine with other methods, such as posting a tweet. 

**Edit:** Resolved in [this post](https://github.com/redouane59/twittered/discussions/385#discussioncomment-3122867).

**Comment by redouane59** [comment]
Can you check [this example](https://github.com/redouane59/twittered/blob/e78aa45a820db1d119eada91051d6fb9b2efd27c/src/test/java/io/github/redouane59/twitter/nrt/ITwitterClientV2AuthenticatedTest.java#L276) and try it on your side ?

**Reply by karsteny** [reply]
Same errors occur with the files. ApiCommands is the name of the class that I am running this in. 
```java
try(InputStream is = ApiCommands.class.getResourceAsStream("/" + "twitter.png");
			ByteArrayOutputStream baos = new ByteArrayOutputStream();){
	byte[] buf = new byte[1024];
	int 	k;
	while ((k = is.read(buf)) > 0) {
		baos.write(buf, 0, k);
	}
	UploadMediaResponse media = client.uploadMedia("twitter.png", baos.toByteArray(), MediaCategory.TWEET_IMAGE);
} catch (IOException e) {
	e.printStackTrace();
}
```

**Reply by redouane59** [reply]
Can you try to run the NRT test on your side ? Because it is passing with success on my side.
Just copy the code and paste it to your project, it should be fine too. :)

**Reply by karsteny** [reply]
Still doesn't work.
I copied all of the V2 tests into the tester, and some of them failed due to the tests not having been updated in a while, and the ImageUpload one still failed, for the exact same reason.
Errors produced by 2 different images (respectively):
```java
javax.net.ssl.SSLProtocolException: Connection reset
Caused by: java.net.SocketException: Connection reset
```
```java
java.io.IOException: Error writing to server
```

Not related, but these tests may need to be fixed as they failed for one reason or another, not too sure if it is my fault or why some happened:
-testPostReplyTweetAndDeleteTweet failed because it is missing the @Username tag that replies have in the getText field
-testFollowAndUnfollow, testCreateListAndAddRemoveMemberAndPinUnpinListAndDeleteList, and testBlockAndUnblockUser failed due to NullPointerExceptions at a test.

**Reply by redouane59** [reply]
For some NRT it's normal that is fails because it depends on the account set up but for the `testPostReplyTweetAndDeleteTweet` it should work. Actually it works on my side. Can you try to debug step by step and see where it crashes ? Can't it be related to your connection blocking some calls ?

**Reply by redouane59** [reply]
You can also check what is the returned messages in the class `AbstractRequestHelper.java` in the variable `stringResponse` of the `makeRequest` method.

**Reply by karsteny** [reply]
What I meant to say about `testPostReplyTweetAndDeleteTweet` is that the test itself is out of date, as whatever `getText()` pulls from the API now has @Username-replying-to in the text.

As for the response in the `stringResponse` method, the execution failed at `makeRequest()`. While doing step-by-step debugging, it would return "Error writing to server" (id=77) as the error, although when running the program at full speed it would throw the SSL exceptions that I originally mentioned or `javax.net.ssl.SSLException: Received fatal alert: record_overflow`. 

I'm not too sure if my connection is blocking the calls. I wouldn't think it is, as I tried uploading the same exact file with Postman (and its built in file management stuff) and it uploaded properly and returned a mediaId. I tried to see if my computer's firewall could be blocking it, but I didn't see anything (although I have quite a limited knowledge of what I was looking at there). 

I'll attach the file I used this attempt here, try it, it shouldn't be out of spec for the API.
![twitter](https://user-images.githubusercontent.com/42128680/178154762-6022e412-31f0-4320-bf64-b4fae3c3905f.png)

**Reply by redouane59** [reply]
Ok let me check it.

**Reply by redouane59** [reply]
Really strange, with my own image (in my test/resources folder), it works well.
![image](https://user-images.githubusercontent.com/4500524/178216543-e7bb8881-9a32-47b3-ad46-45b175f44618.png)

I'm going to try with yours !

**Reply by redouane59** [reply]
Actually, it works also with your image : 
![image](https://user-images.githubusercontent.com/4500524/178217003-87ffc380-df77-435d-b52d-539b32839d91.png)

**Reply by redouane59** [reply]
Can you tell me if other post endpoints work fine ? Like are you able to post a tweet without images ?

**Reply by redouane59** [reply]
`testPostReplyTweetAndDeleteTweet` is also failing on my side, I'm going to check it.

Edit : finally it works, it was just an account permission issue, I just authorised the read & write and now it's ok, you can see the result below : 
![image](https://user-images.githubusercontent.com/4500524/178223964-2b75a169-e5e4-4d77-beee-20eda44fab70.png)

![image](https://user-images.githubusercontent.com/4500524/178224006-ca5f9c87-f3a6-48f7-a5cb-79acc75e8a6e.png)

**Reply by karsteny** [reply]
Alright, I have actually resolved the issue myself.
I googled the error `javax.net.ssl.SSLException: Received fatal alert: record_overflow` to see what exactly could be and [found someone on StackOverflow](https://stackoverflow.com/questions/54119613/sslhandshakeexception-received-fatal-alert-record-overflow) asking about a similar error but with the same details, and someone replied saying it was a bug with Java 11 and Java 12, and that it was fixed in 11.0.5 or 13. So I downloaded a 11.15 JDK and ran the program with that, and lo and behold, it posted just fine. What a strange bug. I guess this question is resolved now, thanks for all the help trying to debug it even though it was something that was wrong with Java itself.

**Reply by redouane59** [reply]
You're welcome :)
